### PR TITLE
Migrate CLI functionality to plugin layer subset 1

### PR DESCRIPTION
**Pull request by cpelley**
First subset of CLI modifications to IMPROVER required by EPP.  This is intended to test the water against the types of change that I would make to the remainder CLI's.

CLI layers modified:
- cli/categorical
  - check_tree was not incorporated into Plugin (looks like a debug usage to me and doesn't return the provided input cubes).
- cli/cloud_condensation_level
- cli/combine
- cli/copy_attributes
- cli/cloud_top_temperature
- cli/standardise

High-level summary of changes:
- New `common_input_handle` utilities: `as_cube` and `as_cubelist`
  - Enable passing either `CubeList` or `Cube` where it makes sense to do so.
  - Allow flexibility to provide Cube/CubeList to plugins.
- New `CopyAttributes` plugin which was before handled by the `copy_attributes` CLI layer.
- New `MetaPluginCloudCondensationLevel` meta-plugin which handles calling `HumidityMixingRatio` and `CloudCondensationLevel`, which `cloud_condensation_level` CLI was doing before.
- Moving cube extraction to the plugin layer.
- Testing captures changes to plugins (like calls to as_cube etc.).
- New plugins are fully tested.

## Issues

- https://github.com/MetOffice/improver_suite/issues/2079
- https://github.com/metoppv/improver/issues/1998
- https://github.com/MetOffice/barebones_workflow/issues/32
- https://github.com/MetOffice/improver_suite/issues/2253

## Mock/patch usage explanation

In this test, we patch both the `as_cube` and `as_cubelist` functions, replacing them with mock objects. This allows us to verify how they are called within the code. Specifically, we define a `side_effect` for the `as_cube` mock object to raise a custom `HaltExecution` exception when invoked. This setup enables us to check that the `as_cube` and `as_cubelist` functions are called with the correct arguments and ensures that the remainder of the plugin code (which isn't under test) does not execute.

The use of `sentinel` helps create unique, identifiable objects for testing purposes. By passing a dummy template cube object (a `sentinel`), we can verify that the mocked `as_cube` function is called with it as its argument. This approach simplifies the test setup by eliminating the need to create real iris cubes or handle the complexities associated with them. Thus, our unit test remains focused and isolated, only testing the behavior under consideration without external dependencies.

```python
class HaltExecution(Exception):
    pass


@patch("improver.utilities.copy_attributes.as_cube")
@patch("improver.utilities.copy_attributes.as_cubelist")
def test_as_cubelist_called(mock_as_cubelist, mock_as_cube):
    mock_as_cubelist.return_value = sentinel.cubelist
    mock_as_cube.side_effect = HaltExecution
    try:
        CopyAttributes(["attribA", "attribB"])(
            sentinel.cube0, sentinel.cube1, template_cube=sentinel.template_cube
        )
    except HaltExecution:
        pass
    mock_as_cubelist.assert_called_once_with(sentinel.cube0, sentinel.cube1)
    mock_as_cube.assert_called_once_with(sentinel.template_cube)
```

### CODE REVIEW

**Comment by cpelley** [comment]
@gavinevans, @bayliffe 

I think we want to eventually potentially move away from the use of the term 'Plugin', so any ideas for suitable alternative naming to `MetaPluginCloudCondensationLevel` are welcome.

I couldn't get the pre-commit environment install to complete (pip install).
Before I chase this up, perhaps you are aware already of a solution?

File: improver/utilities/copy_attributes.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        Args:
+            attributes:
+                List of names of attributes to copy. If any are not present on template_cube, a
+                KeyError will be raised.
+        """
+        self.attributes = attributes
+
+    def process(
+        self, *cubes: Union[Cube, CubeList], template_cube: Union[Cube, CubeList]
+    ) -> Union[Tuple[Union[Cube, CubeList]], Cube, CubeList]:
```
Comment (cpelley) [inline improver/utilities/copy_attributes.py:30]: Notice how for plugins that can accept any number of inputs, I would propose this return behaviour.
That being, return a Cube or CubeList, if only 1 input was provided, otherwise, return a tuple of Cube/CubeList objects corresponding to our inputs arguments.

File: improver_tests/categorical/decision_tree/test_ApplyDecisionTree.py [hunk 2]
```diff
@@ -56,6 +57,22 @@
 from . import deterministic_diagnostic_tree, wxcode_decision_tree
 
 
+class HaltExecution(Exception):
+    pass
+
+
+@patch("improver.categorical.decision_tree.as_cubelist")
```
Comment (cpelley) [inline improver_tests/categorical/decision_tree/test_ApplyDecisionTree.py:64]: We ensure that each plugin utilises the standardised input handling approach.

File: improver_tests/psychrometric_calculations/cloud_condensation_level/test_MetaPluginCloudCondensationLevel.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+
+@patch(
+    "improver.psychrometric_calculations.psychrometric_calculations.HumidityMixingRatio"
+)
+@patch(
+    "improver.psychrometric_calculations.cloud_condensation_level.CloudCondensationLevel"
+)
+def test_basic(mock_CloudCondensationLevel, mock_HumidityMixingRatio):
```
Comment (cpelley) [inline improver_tests/psychrometric_calculations/cloud_condensation_level/test_MetaPluginCloudCondensationLevel.py:18]: This simply ensures that this meta-plugin, calls the underlying plugins in the way we intended.

File: improver_tests/utilities/copy_attributes/test_CopyAttributes.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    plugin = CopyAttributes(attributes)
+    result = plugin.process(cube0, cube1, template_cube=template_cube)
+    assert type(result) == tuple
+    for res in result:
+        assert res.attributes["attribA"] == "tempA"
+        assert res.attributes["attribB"] == "tempB"
+        assert "attribC" not in res.attributes
+    assert id(result[0]) == id(cube0)
+    assert id(result[1]) == id(cube1)
```
Comment (cpelley) [inline improver_tests/utilities/copy_attributes/test_CopyAttributes.py:54]: Note how we are verifying the object IDs, since `CubeAttributes` is an in-place operation.

File: improver/cli/standardise.py [hunk 5]
```diff
@@ -81,13 +81,8 @@ def process(
     Returns:
         iris.cube.Cube
     """
-    from improver.metadata.amend import update_stage_v110_metadata
     from improver.standardise import StandardiseMetadata
 
-    # update_stage_v110_metadata is deprecated. Please ensure metadata is
-    # StaGE version 1.2.0 compatible.
-    update_stage_v110_metadata(cube)
```
Comment (cpelley) [inline improver/cli/standardise.py:89]: This was removed as per https://github.com/metoppv/improver/pull/1986

File: improver_tests/categorical/decision_tree/test_ApplyDecisionTree.py [hunk 6]
```diff
@@ -56,6 +57,22 @@
...
+def test_as_cubelist_called(mock_as_cubelist):
+    mock_as_cubelist.side_effect = HaltExecution
+    try:
+        ApplyDecisionTree(decision_tree=deterministic_diagnostic_tree())(
+            sentinel.precip_cube, sentinel.hail_cube
+        )
+    except HaltExecution:
+        pass
+    mock_as_cubelist.assert_called_once_with(sentinel.precip_cube, sentinel.hail_cube)
+
```
Comment (bayliffe) [inline improver_tests/categorical/decision_tree/test_ApplyDecisionTree.py:74]: I don't understand any of this I'm afraid. I may need some pointers on mock, and whatever sentinel is, at some point.

File: improver_tests/cube_combiner/test_Combine.py [hunk 7]
```diff
@@ -146,5 +161,5 @@ def test_minimum_realizations_exceptions(
 
 def test_empty_cubelist():
     """Ensure supplying an empty CubeList raises an error"""
-    with pytest.raises(TypeError, match="A cube is needed to be combined."):
+    with pytest.raises(ValueError, match="A cube is needed to be combined."):
```
Comment (bayliffe) [inline improver_tests/cube_combiner/test_Combine.py:164]: Ditch this if you remove this test from the plugin and rely on the as_cubelist function to raise a suitable exception.

File: improver_tests/utilities/common_input_handle/test_as_cube.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+import pytest
+from iris.cube import Cube, CubeList
+
+from improver.utilities.common_input_handle import as_cube
+
+
+def test_cubelist_as_cube():
```
Comment (bayliffe) [inline improver_tests/utilities/common_input_handle/test_as_cube.py:37]: I know they're very simple tests, but doc-strings give a lovely feel of quality to a file. Could you lavish some on these tests?

File: improver_tests/utilities/common_input_handle/test_as_cube.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+def test_no_cube_provided():
+    msg = "A cube should be provided."
+    with pytest.raises(ValueError, match=msg):
+        as_cube(None)
+
+
+def test_non_cube_provided():
+    msg = "A cube should be provided."
+    with pytest.raises(TypeError, match=msg):
+        as_cube(["cube"])
```
Comment (bayliffe) [inline improver_tests/utilities/common_input_handle/test_as_cube.py:59]:
```suggestion
        as_cube(CubeList(["cube"]))
```
If you modified this test as above it would presently return the word "cube" without complaint.

File: improver/utilities/common_input_handle.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    Returns:
+        Cube:
+            A single cube.
+    """
+    if not cube:
+        raise ValueError("A cube should be provided.")
+    if isinstance(cube, CubeList):
+        if len(cube) > 1:
+            raise ValueError("A single cube should be provided.")
+        cube = cube[0]
```
Comment (bayliffe) [inline improver/utilities/common_input_handle.py:77]: There is no checking here of what is contained within the cube list (see the unit tests where I highlight how this would return a non-cube).

File: improver_tests/utilities/common_input_handle/test_as_cubelist.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+# POSSIBILITY OF SUCH DAMAGE.
+import pytest
+from iris.cube import Cube, CubeList
+
+from improver.utilities.common_input_handle import as_cubelist
+
+
+def test_cubelist_as_cubelist():
+    cube = Cube([0])
+    res = as_cubelist(cube)
```
Comment (bayliffe) [inline improver_tests/utilities/common_input_handle/test_as_cubelist.py:39]: I think you meant to pass in a cubelist here, rather than a cube; currently it is identical to the test below.
(Wouldn't a lovely doc-string help make that clearer? Go on, you know you want some lovely doc-strings).

File: improver_tests/utilities/common_input_handle/test_as_cubelist.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    assert id(res[0]) == id(cube)
+
+
+def test_cube_cubelist_mixture_as_cubelist():
+    cube = Cube([0])
+    cubes = CubeList([cube])
+    res = as_cubelist(cube, cubes)
+    assert isinstance(res, CubeList)
+    assert id(res[0]) == id(cube)
+    assert id(res[1]) == id(cube)
```
Comment (bayliffe) [inline improver_tests/utilities/common_input_handle/test_as_cubelist.py:57]: Can you please create two distinct cubes for this test. It tests nothing of whether the returned order is as expected as the same cube is returned in both positions.

File: improver/utilities/common_input_handle.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        cubes:
+            Input data provided in the form of one or more cubes or cubelists (or mixture thereof).
+
+    Returns:
+        CubeList:
+            A CubeList containing all the cubes provided as input.
+    """
+    if not cubes or all([not cube for cube in cubes]):
+        raise ValueError("One or more cube should be provided.")
+    return CubeList(flatten(cubes))
```
Comment (bayliffe) [inline improver/utilities/common_input_handle.py:55]: There is no testing of what is contained within the cubelist here at all.
You should check, after flattening to simplify things, that every element of the final cube list is of Cube type.
This will need a unit test.

File: improver_tests/utilities/common_input_handle/test_as_cubelist.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+
```
Comment (bayliffe) [inline improver_tests/utilities/common_input_handle/test_as_cubelist.py:59]:
```
def test_cubelist_mania():
    cube1 = "kittens"
    cube2 = "aren't"
    cube3 = "cubes"
    cubes1 = CubeList([cube1, cube2])
    res = as_cubelist(cubes1, cube3)
    assert isinstance(res, CubeList)
    assert id(res[0]) == id(cube1)
    assert id(res[1]) == id(cube2)
    assert id(res[2]) == id(cube3)
```
This test also passes happily at the moment. Awww, kittens.

File: improver_tests/utilities/copy_attributes/test_CopyAttributes.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    mock_as_cube.assert_called_once_with(sentinel.template_cube)
+
+
+def test_copy_attributes_multi_input():
+    """
+    Test the copy_attributes function for multiple input cubes.
+
+    Demonstrates copying attributes from the template cube to the input
+    cubes and also demonstrates the attributes on the templates cube that
+    aren't specified in the attributes list awre indeed ignored.
```
Comment (bayliffe) [inline improver_tests/utilities/copy_attributes/test_CopyAttributes.py:63]:
```suggestion
    aren't specified in the attributes list are indeed ignored.
```

File: improver_tests/utilities/copy_attributes/test_CopyAttributes.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    assert id(result[0]) == id(cube0)
+    assert id(result[1]) == id(cube1)
+
+
+def test_copy_attributes_single_input():
+    """
+    As per 'test_copy_attributes_multi_input' except only one input cube is provided.
+    """
+    attributes = ["attribA", "attribB"]
+    cube0 = Cube([0], attributes={"attribA": "valueA", "attribB": "valueB"})
```
Comment (bayliffe) [inline improver_tests/utilities/copy_attributes/test_CopyAttributes.py:88]: Can you please modify these values not to match the template cube so the impact of apply the plugin can be seen.
Can you please add an attribute to cube0 that is not on the template cube (attribD) to demonstrate that it is left in place and unchanged by the application of the plugin.

**Review by bayliffe — DISMISSED** [review]
Have a bucket of comments. You lucky thing.

File: improver/cube_combiner.py [hunk 17]
```diff
@@ -128,8 +129,11 @@ def process(self, cubes: CubeList) -> Cube:
             ValueError:
                 If minimum_realizations aren't met, or less than one were requested.
         """
-        if not cubes:
-            raise TypeError("A cube is needed to be combined.")
+        try:
+            cubes = as_cubelist(*cubes)
+        except ValueError:
+            raise ValueError("A cube is needed to be combined.")
```
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/cube_combiner.py:135]: This was left here as this particular exception msg was raised originally.
Happy to remove.

File: improver/utilities/copy_attributes.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        Operation is performed in-place on provided inputs.
+
+        Args:
+            cubes:
+                Source cube(s) to be updated.
+            template_cube:
+                Source cube to get attribute values from.
+
+        Raises:
+            KeyError: If any of the attributes are not present on the template_cube.
```
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/utilities/copy_attributes.py:69]: True.  The dilemma is that if not here, people would otherwise have to trace through the source code to determine what is done to `template_cube` to determine whether the `amend_attributes` function would even raise an exception (and that's us pointing people to the amend_attributes function from the docstring of this class).  On the other side, it is `amend_attributes` that raises the exception so documenting here is at risk of becoming outdated.

...I'll remove

File: improver_tests/cube_combiner/test_Combine.py (hunk 7 above)
Comment (cpelley) [reply improver_tests/cube_combiner/test_Combine.py:164]: Sure, I assumed to maintain non-functional changes where it made sense to do so.
As you have now confirmed, happy to remove (as with anything where I have avoided changing things) 👍

**Comment by cpelley** [comment]
Thanks for your thorough review @bayliffe ❤️ 
I'll take action now.  I'll also update the PR description with details around the use of mock, sentinel etc. (apologies for this oversight).

**Comment by TomekTrzeciak** [comment]
> I think we want might eventually want to move away from the use of the term 'Plugin',

Hallelujah!

File: improver/utilities/common_input_handle.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+        cube:
+            Input data provided in the form of a cube ir cubelists.
+
+    Returns:
+        Cube:
+            A single cube.
+    """
+    if not cube:
+        raise ValueError("A cube should be provided.")
+    if isinstance(cube, CubeList):
```
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/utilities/common_input_handle.py:74]: if not a cube or cubelist an exception is raised to say that a cube must be provided

File: improver/psychrometric_calculations/cloud_condensation_level.py [hunk 20]
```diff
@@ -45,6 +45,53 @@
...
+        Args:
+            cubes:
+                Cubes, of temperature (K), pressure (Pa) and humidity (1).
+
+        Returns:
+            Cubes of air_temperature_at_cloud_condensation_level and
+            air_pressure_at_cloud_condensation_level
+
+        """
+        humidity = self._humidity_plugin(*cubes)
```
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/psychrometric_calculations/cloud_condensation_level.py:91]: The HumidityMixingRatio plugin is what handles the cube input.  It's process method was updated to perform a cube extract:

https://github.com/metoppv/improver/blob/562ee522616d0da9e2dcab8ebb932428c85a43ec/improver/psychrometric_calculations/psychrometric_calculations.py#L357-L364

File: improver/cli/copy_attributes.py [hunk 21]
```diff
@@ -57,8 +57,7 @@ def process(
     Returns:
         iris.cube.Cube
     """
-    from improver.metadata.amend import amend_attributes
+    from improver.utilities.copy_attributes import CopyAttributes
```
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/cli/copy_attributes.py:60]: The CLI layer UI remains unchanged, expecting only 1 cube and 1 template

File: improver/utilities/common_input_handle.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/common_input_handle.py:74]: I was wondering if you wanted to handle `[cube1, cube2]`, as well as `iris.cube.CubeList([cube1, cube2])`.

File: improver/utilities/common_input_handle.py (hunk 13 above)
Comment (bayliffe) [reply improver/utilities/common_input_handle.py:55]: Sure....
```
a = iris.cube.CubeList(["hello", "there"])
print(a[0])
>>> hello
```

File: improver_tests/utilities/common_input_handle/test_as_cube.py (hunk 9 above)
Comment (cpelley) [reply improver_tests/utilities/common_input_handle/test_as_cube.py:59]: Ah the version of iris we use is old enough that the cubelist containing cubes check in iris wasn't added yet :(

File: improver/cli/copy_attributes.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/cli/copy_attributes.py:60]: https://github.com/metoppv/improver/pull/2000/files#diff-f9123f095bd65dd049d587a670016b72151441287826b8ecd760055a0bc47ee6R78

This line in CopyAttributes is new (as that whole plugin is new). It loops over target cubes and allows the copying of the attributes from one source cube to these targets (multiple targets).

Are you suggesting that is not accessible via the CLI for some reason? If it is, the doc-string should reflect that.

File: improver/psychrometric_calculations/cloud_condensation_level.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/psychrometric_calculations/cloud_condensation_level.py:91]: Yeah, not sure what confused me here, possibly looking at the old version at the same time and missed the extract in the plugin.

File: improver/utilities/common_input_handle.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/utilities/common_input_handle.py:79]: well `as_cube` would currently throw an exception for the case where a cubelist is provided containing more than 1 cube (ultimately one could extend to internally call merge_cube perhaps or provide a UI to do this).

I saw `as_cube` as taking the role of returning a single cube (raising an exception where this isn't possible).
I saw `as_cubes` as taking the role of returning a single cubelist.

Motivated by the fact that I think that any function or class should needn't hardcode an expectation to be provided a cube, since a cubelist containing a single cube should be equally suitable.

Retuning more than 1 cube, well I think that's the job of the cubelist, hence why I thought `as_cube` would be best placed returning only the single cube case.  That is, I think `as_cubes([cube1, cube2])` would be better to use no??

**Comment by cpelley** [comment]
@bayliffe, I have updated the ticket description with an explanation of my usage of mock/path/sentinel.

**Comment by cpelley** [comment]
I have made changes based on feedback.  For now I have made it so that `as_cubes` accepts ONLY a `Cube`/`CubeList` (can add another time??).
Additionally, since I was using flatten a great deal, I have re-implemented it to be what I feel is more robust and flexible.
I have merged in master and updated the headers of the files added accordingly.
Cheers,

**Comment by cpelley** [comment]
@bayliffe, appologies for making additional changes.  I looked back at this and felt that perhaps what I had done wasn't quite aligned with your review feedback.  Now, in summary:

- `as_cubelist` can be passed pretty much anything and it just verifies that a cubelist containing cubes is returned.
- `as_cube` now calls `as_cubelist` under the hood, then either returns the first index (when of length 1) or otherwise performs a `CubeList.merge_cube` in an attempt to return a single cube.

File: improver/psychrometric_calculations/cloud_top_temperature.py [hunk 22]
```diff
@@ -112,23 +114,45 @@ def _make_cct_cube(self, data: ndarray) -> Cube:
...
+        )
+        self.p_at_ccl = cubes.extract(
+            [
+                "air_pressure_at_condensation_level",
+                "air_pressure_at_cloud_condensation_level",
+            ]
+        )
+        self.temperature = cubes.extract(
+            ["temperature_on_pressure_levels", "air_temperature"]
+        )
```
Comment (cpelley) [inline improver/psychrometric_calculations/cloud_top_temperature.py:143]: These alternate names were provided in order to ensure that the tests would still succeed.
Perhaps the tests don't have the correct names and need updating instead of this flexible extract?

Alternative, if we need to be more relaxed about naming here, we can do something like I see you have done elsewhere.  That is, extract all the cube names, then match against the names containing:

- "air_temperature"
- "air_pressure"
- remaining cube containing "temperature"

I'm all 👂's

File: improver/psychrometric_calculations/psychrometric_calculations.py [hunk 23]
```diff
@@ -314,20 +315,31 @@ def _make_humidity_cube(self, data: np.ndarray) -> Cube:
...
-        self.temperature, self.pressure, self.rel_humidity = cubes
+        cubes = as_cubelist(*cubes)
+        try:
+            (self.temperature, self.pressure, self.rel_humidity,) = cubes.extract(
+                ["air_temperature", "surface_air_pressure", "relative_humidity"]
+            )
+        except ValueError:
+            (self.temperature, self.pressure, self.rel_humidity,) = cubes.extract(
+                ["air_temperature", "air_pressure", "relative_humidity"]
+            )
```
Comment (cpelley) [inline improver/psychrometric_calculations/psychrometric_calculations.py:338]: This is again due to unittests using a different name to the CLI tested

**Comment by cpelley** [comment]
The second half of EPP required IMPROVER plugin changes are captured in https://github.com/metoppv/improver/pull/2003 (that PR builds on this for the remaining plugins).

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks @cpelley 👍 

I've added some minor comments.

File: improver/cli/categorical.py [hunk 24]
```diff
@@ -62,21 +62,14 @@ def process(
     if check_tree:
         from improver.categorical.utilities import check_tree
 
+        # WARNING: This means that we do not return our input data
```
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/cli/categorical.py:65]: Well `check_tree` returns a [string](https://github.com/metoppv/improver/blob/a989fed040fa942be10e9f9a8e320398920730c2/improver/categorical/utilities.py#L556).  I think this is a debug call of the cli which I wouldn't expect to be used within our graph (we don't use it at least).
Perhaps I should instead say, "This means that we do not return a cube/cubelist"  (I think that's the significant issue here)

File: improver/utilities/common_input_handle.py [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+            Any iterable is supported.
+
+    Returns:
+        CubeList:
+            A CubeList containing all the cubes provided as input.
+    """
+    cubes = CubeList(flatten(cubes))
+    # Remove CubeList verification for iris >=3.3.0
+    for cube in cubes:
+        if not hasattr(cube, "add_aux_coord"):
```
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/utilities/common_input_handle.py:31]: In recent versions of iris, `hasattr` is the approach taken within the CubeList cube check built-in:
https://github.com/SciTools/iris/blob/b15a1f61b7d48ca9c7af464c5a5f112ac7366bc7/lib/iris/cube.py#L154-L158

When we upgrade to newer iris, we will be subject to iris CubeList built-in behaviour anyway.  I have preferenced not deviating from later iris version behaviour to avoid hiding potential issues that may arise later from the upgrade.

As an aside, the 'pythonic' approach is indeed to generally avoid `isinstance` calls where possible.  The idea is that if something has the interface you need, then it doesn't matter what the object actually is (e.g. code handling numpy/dask arrays).
However, in this case, I actually very much agree with you that an `isinstance` call is preferable in iris CubeList cube checking :(
(cubelist behaviour is not limited to an add_aux_coord interface of the object and so is not indication enough of the validity of the objects it contains)

File: improver/psychrometric_calculations/cloud_condensation_level.py [hunk 26]
```diff
@@ -19,6 +19,52 @@
...
+        """
+        Set up class
+
+        Args:
+            model_id_attr:
+                Name of model ID attribute to be copied from source cubes to output cube
+        """
+        from improver.psychrometric_calculations.psychrometric_calculations import (
+            HumidityMixingRatio,
+        )
```
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/psychrometric_calculations/cloud_condensation_level.py:43]: The motivation was in enforcing that a science configuration would have to reference a plugin from the module where it is defined, rather than from some other module that uses it.  I'll move the import to the top at this time.  Can peruse this potential issue in future.

File: improver/psychrometric_calculations/cloud_top_temperature.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/psychrometric_calculations/cloud_top_temperature.py:143]: Thanks for confirming 👍

File: improver/psychrometric_calculations/psychrometric_calculations.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/psychrometric_calculations/psychrometric_calculations.py:338]: I'll correct the unittests 👍

**Comment by cpelley** [comment]
Thanks @bayliffe, @gavinevans 
I have made changes in reflection of the reviews given 👍

File: improver/utilities/common_input_handle.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (gavinevans) [reply improver/utilities/common_input_handle.py:31]: OK, I wouldn't have expected that checking for this attribute would be how iris would identify cubes or cubelists internally. As you say, we may as well adhere to this form of checking given that this is what is done in the newer versions of Iris. I hadn't appreciated that checking for an attribute in this way would be considered more "pythonic" but I understand that just checking for the existence of a particular attribute, if that is your only concern, would be more flexible. I'm not sure that raising this as an iris issue will necessarily be worthwhile, so I'm happy for you to leave this.

File: improver_tests/utilities/common_input_handle/test_as_cube.py [hunk 27]
```diff
@@ -0,0 +1,53 @@
+# (C) Crown copyright, Met Office. All rights reserved.
+#
+# This file is part of IMPROVER and is released under a BSD 3-Clause license.
+# See LICENSE in the root of the repository for full licensing details.
+from unittest.mock import patch, sentinel
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply improver_tests/utilities/common_input_handle/test_as_cube.py:6]: Yes, these docstrings are a bit superfluous, although I'm a bit reluctant to change our conventions unilaterally in this PR without some discussion/agreement in the team, so as to avoid confusion in other PRs.

File: improver/psychrometric_calculations/cloud_condensation_level.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (gavinevans) [reply improver/psychrometric_calculations/cloud_condensation_level.py]: OK, I wouldn't have considered this possibility. Is this your motivation for the change in `test_MetaPluginCloudCondensationLevel.py` to use 
```
@patch(
    "improver.psychrometric_calculations.cloud_condensation_level.HumidityMixingRatio"
)
```
rather than 
```
@patch(
    "improver.psychrometric_calculations.psychrometric_calculations.HumidityMixingRatio"
)
```
even though the `HumidityMixingRatio` plugin is defined within psychrometric_calculations.py?

Why would someone do this?

File: improver_tests/utilities/common_input_handle/test_as_cube.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (cpelley) [reply improver_tests/utilities/common_input_handle/test_as_cube.py:6]: https://github.com/metoppv/improver/issues/2004

File: improver/psychrometric_calculations/cloud_condensation_level.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (cpelley) [reply improver/psychrometric_calculations/cloud_condensation_level.py]: > Is this your motivation for the change in test_MetaPluginCloudCondensationLevel.py

When it comes to patching, that is a different matter (it is the `MetaPluginCloudCondensationLevel` class under test which requires its objects from the cloud_condensation_level to be patched).

My concern was more around a science configuration specifically.  If someone references a plugin from a module that uses it, to execute, then that creates an implicit dependency on this other module.  That is, if someone changes that module, it could break their configuration even though they aren't intending to make changes to the plugin itself from the module where it is defined.
This opens up a wider discussion around discoverability etc. which is likely beyond the scope of this PR.  CLIs are all available in a flat space namespace.  That is, all under `improver.cli`
Similarly, I think it might be useful to make plugins available under a flat namespace (thereby protecting science configurations from changes in project layout/structure).  For example, perhaps `improver.plugin.MetaPluginCloudCondensationLevel`.
As I say, I think this is beyond the scope of this PR and warrants its own PR so as not to detract from the changes of this work.

**Comment by cpelley** [comment]
Thanks @gavinevans, I think all feedback has been actioned.

I'm aware that have yet to take the bait on discussing naming convention to these meta 'plugins' (https://github.com/metoppv/improver/pull/2000#issuecomment-2120557385).
In the second part of changes to IMPROVER CLI/plugins work, I opted for calling these meta 'plugins' `MetaProcMod<name>`
See [here](https://github.com/metoppv/improver/pull/2003/files#diff-3d2a4fb489eed048fde2ad728d30e298d9ebaf98d421f7fba1d86102166294fc)
The motivation is around moving away from the what might be deemed a term that implies what it is not, to the more generic 'processing module' term.  The later does not share the same technical implications of a technical term 'plugin'.
Would you be content for me to rename `MetaPluginCloudCondensationLevel` -> `MetaProcModCloudCondensationLevel`?

New issues (out of scope of this PR):

- https://github.com/metoppv/improver/issues/2005
- https://github.com/metoppv/improver/issues/2004

File: improver/psychrometric_calculations/cloud_condensation_level.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (gavinevans) [reply improver/psychrometric_calculations/cloud_condensation_level.py]: OK, I can understand the concern from a technical perspective where this kind of implicit dependency could be created. In practice, I'm not aware that we ever import things like that to create an implicit dependency, but I can see that it would be possible. I think that your comments about plugin discoverability are interesting (#2005) and probably a good idea.

**Comment by gavinevans** [comment]
@cpelley I'm not sure that I'm best placed to comment on the use of the word "plugin" to describe the IMPROVER "plugins", but if you'd like to not include this term within your `Meta` "plugins", then that seems fine from my perspective.

**Review by gavinevans — DISMISSED** [review]
Thanks for these updates, @cpelley 👍

File: improver/utilities/flatten.py [hunk 28]
```diff
@@ -4,27 +4,27 @@
...
+    ):
+        yield nested_iterable
+        return
+    for item in nested_iterable:
+        if isinstance(item, Iterable) and not isinstance(item, (str, bytes)):
+            yield from flatten(item)
         else:
-            flat_list.append(item)
-    return flat_list
+            yield item
```
Comment (bayliffe) [inline improver/utilities/flatten.py:30]: I was surprised that this fundamental change to the return type from a list to a generator had few knock on effects. The returned object can only be used once, it cannot be indexed, etc. without conversion to a list. I note that you've wrapped the output in a `list()` in most places, so that explains that.

The only comment I have is that the doc-string here:
https://github.com/metoppv/improver/blob/dee4f9b1ac20b6c3adfab0d124fdc5be54b6e344/improver/utilities/forecast_reference_enforcement.py#L333
ought to reflect that `cubes` could be a generator, given that the only place this actually gets called (normalise_to_reference) now passes in a generator returned by the `flatten` function.

File: improver_tests/utilities/common_input_handle/test_as_cubelist.py [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+    msg = "One or more cubes should be provided."
+    with pytest.raises(ValueError, match=msg):
+        as_cubelist([])
+
+
+def test_non_cube_cubelist_provided():
+    """Test when a CubeList containing a non cube is provided."""
+    msg = "A non iris Cube object has been provided."
+    with pytest.raises(TypeError, match=msg):
+        as_cubelist(CubeList(["not_a_cube"]))
```
Comment (bayliffe) [inline improver_tests/utilities/common_input_handle/test_as_cubelist.py:79]: I can't believe you didn't get any kittens into these tests. Are you even trying?

File: improver_tests/utilities/test_flatten.py [hunk 30]
```diff
@@ -8,48 +8,56 @@
...
+        # Edge cases
+        ([], []),
+        ([1], [1]),
+        (iter([]), []),
+        ({frozenset([])}, []),
+        # Non-iterables - just yield the item, don't attempt to flatten
+        (42, [42]),
+        ("string", ["string"]),
+        (b"bytes", [b"bytes"]),
+    ],
```
Comment (bayliffe) [inline improver_tests/utilities/test_flatten.py:54]: This is a nice improvement, thanks.

**Review by bayliffe — APPROVED** [review]
I've added one comment about a doc-string, but I am not enormously worried about it. If you make a change, nudge me and I'll reapprove.

**Review by bayliffe — APPROVED** [review]
